ext/flash/socket.rs

```diff
@@ -0,0 +1,135 @@
+use mio::net::TcpStream;
+use std::{
+  cell::UnsafeCell,
+  io::{Read, Write},
+  sync::{Arc, Mutex},
+};
+use tokio::{io::AsyncWrite, sync::mpsc};
+
+use crate::ParseStatus;
+
+type TlsTcpStream = rustls::StreamOwned<rustls::ServerConnection, TcpStream>;
+
+pub enum InnerStream {
+  Tcp(TcpStream),
+  Tls(Box<TlsTcpStream>),
+}
+
+pub struct Stream {
+  pub inner: InnerStream,
+  pub detached: bool,
+  pub read_rx: Option<mpsc::Receiver<()>>,
+  pub read_tx: Option<mpsc::Sender<()>>,
+  pub parse_done: ParseStatus,
+  pub buffer: UnsafeCell<Vec<u8>>,
+  pub read_lock: Arc<Mutex<()>>,
+}
+
+impl Stream {
+  pub fn detach_ownership(&mut self) {
+    self.detached = true;
+  }
+
+  pub fn reattach_ownership(&mut self) {
+    self.detached = false;
+  }
+
+  /// Try to write to the socket. If socket will block, return the amount of bytes left.
+  #[inline]
+  pub fn try_write(&mut self, buf: &[u8]) -> usize {
+    let mut nwritten = 0;
+    while nwritten < buf.len() {
+      match self.write(&buf[nwritten..]) {
+        Ok(n) => nwritten += n,
+        Err(ref e) if e.kind() == std::io::ErrorKind::WouldBlock => {
+          break;
+        }
+        Err(e) => {
+          log::trace!("Error writing to socket: {}", e);
+          break;
+        }
```

Comment:
You should probably return the error if nwritten == 0.

Reply:
`try_write` is called in the Fast API op, JS will retry the async path (which can throw errors) if this returns 0